security/pkg/nodeagent/vm/nodeagent.go

```diff
@@ -59,26 +64,70 @@ func (na *nodeAgentInternal) Start() error {
 	na.identity = identity
 	var success bool
 	for {
-		privateKey, req, reqErr := na.createRequest()
-		if reqErr != nil {
-			return reqErr
+		var csrResponseFromOldAPI *pb.CsrResponse
+		var keyPEM []byte
+		var csrPEM []byte
+		var certChainPEM []string
+		var csrError error
+		if na.config.CAClientConfig.IsCitadel {
+			privateKey, req, reqErr := na.createRequest()
+			if reqErr != nil {
+				return reqErr
+			}
+			csrResponseFromOldAPI, csrError = na.caProtocol.SendCSR(req)
+			keyPEM = privateKey
+		} else {
+			options := pkiutil.CertOptions{
+				Host:       identity,
+				Org:        na.config.CAClientConfig.Org,
+				RSAKeySize: na.config.CAClientConfig.RSAKeySize,
+				IsDualUse:  na.config.DualUse,
+			}
+
+			csrPEM, keyPEM, err = pkiutil.GenCSR(options)
+			if err != nil {
+				log.Errorf("Failed to generated key cert for %q: %v", identity, err)
+			}
+			jwt, err := na.pc.GetAgentCredential()
+			if err != nil {
+				log.Errorf("Error getting JWT from metadata server: %v", err)
+			}
+			stsClient := stsclient.NewPlugin()
+			// TODO(pitlv2109): Periodically update the JWT and OAuth2 token.
+			// TODO(pitlv2109): Extract the project_id field from the JWT.
+			token, _, _ := stsClient.ExchangeToken(context.Background(), "csm-fsa@mesh-expansion-235023.iam.gserviceaccount.com", string(jwt[:]))
+			certChainPEM, csrError = na.caClient.CSRSign(context.Background(), csrPEM, token, int64(na.config.CAClientConfig.RequestedCertTTL.Minutes()))
 		}
 
 		log.Infof("Sending CSR (retrial #%d) ...", retries)
 
-		resp, err := na.caProtocol.SendCSR(req)
-		if err == nil && resp != nil && resp.IsApproved {
-			waitTime, ttlErr := na.certUtil.GetWaitTime(resp.SignedCert, time.Now())
+		if na.isCsrRequestApproved(csrResponseFromOldAPI, csrError) {
+			var waitTime time.Duration
+			var ttlErr error
+			if na.config.CAClientConfig.IsCitadel {
+				waitTime, ttlErr = na.certUtil.GetWaitTime(csrResponseFromOldAPI.SignedCert, time.Now())
+			} else {
+				waitTime, ttlErr = na.certUtil.GetWaitTime([]byte(certChainPEM[0]), time.Now())
+			}
 			if ttlErr != nil {
 				log.Errorf("Error getting TTL from approved cert: %v", ttlErr)
 				success = false
 			} else {
+				var certChain []byte
+				if na.config.CAClientConfig.IsCitadel {
+					certChain = append(csrResponseFromOldAPI.SignedCert, csrResponseFromOldAPI.CertChain...)
+				} else {
+					for _, c := range certChainPEM {
+						certChain = append(certChain, []byte(c)...)
+					}
+				}
 				if err = caclient.SaveKeyCert(na.config.CAClientConfig.KeyFile,
 					na.config.CAClientConfig.CertChainFile,
-					privateKey, append(resp.SignedCert, resp.CertChain...)); err != nil {
+					keyPEM, certChain); err != nil {
 					return err
 				}
 				log.Infof("CSR is approved successfully. Will renew cert in %s", waitTime.String())
+				fmt.Println(certChainPEM)
```

Comment:
Remove this before merging.

Reply:
Or change to Debug

Reply:
Absolutely. This is just for testing purposes :)